### ISSUE TITLE
1.14.0 breaks write_string semantics

### ISSUE THREAD

**Issue by skaes** [ffi/ffi#857]
I find the change that automatically adds a null byte to the target of a write_string call somewhat problematic.

Previously, write_string(s, n) wrote exactly n bytes to the destination, now it writes n+1 bytes. This has the potential to break quite a few programs in horrible ways, especially those who allocate a buffer of s.byte_size before calling write_string(s, s.byte_size) on it. One example can be found here: https://github.com/chuckremes/ffi-rzmq/blob/master/lib/ffi-rzmq/message.rb#L126-L135

I'm wondering whether it would be better to introduce a new method with the new semantics and leave the existing method as is.

**Comment by eregon** [comment]
You can use `write_bytes` to avoid the final null byte.
In any case, this seems worth fixing in `ffi-rzmq`.

If the size of the FFI::Pointer is known, you'll get a nice exception telling you to use `write_bytes` instead.

But https://github.com/chuckremes/ffi-rzmq/blob/740a413ce5a13104fb06702bade68f86fa7626b2/lib/ffi-rzmq/message.rb#L126-L135 calls `malloc()` directly, so FFI doesn't know the size of the returned pointer
(although I guess potentially FFI could treat malloc specially and give the passed argument as size of the pointer).

Some background behind the change in https://github.com/ffi/ffi/issues/805 and https://github.com/ffi/ffi/pull/806.

**Comment by skaes** [comment]
@eregon I already created a PR for ffi-rzmq. That wasn't really why I raised the issue. I guess I had assumed that such a breaking change would be something to introduce with a 2.x release, if at all.

**Comment by eregon** [comment]
The usage linked in `ffi-rzmq` is:
https://github.com/chuckremes/ffi-rzmq/blob/740a413ce5a13104fb06702bade68f86fa7626b2/lib/ffi-rzmq/message.rb#L129
```ruby
      data_buffer.write_string bytes, len
```

I also found two similar usages of `write_string_length` in OptCarrot:
https://github.com/mame/optcarrot/blob/5d7a56e13ae69b4739ae83f28a76aff311baedf7/lib/optcarrot/driver/sdl2_audio.rb#L52-L55
```ruby
        stream.write_string_length(@buff, buff_size)
        @buff.clear
      else
        stream.write_string_length(@buff, stream_len)
```

All 3 usages clearly want to write `len` bytes and nothing more.
My initial PR was not null-terminating when `len` is given (https://github.com/ffi/ffi/pull/806#issuecomment-660683601).

So, what I think we can do here is to check if the passed `string.bytesize == len`, if so, warn and use `write_bytes` semantics.
So we'd extend this condition:
https://github.com/ffi/ffi/blob/0d8e5815716eeeb8fad11205646dfedd8bf23f03/lib/ffi/pointer.rb#L104
to:
`if len == size or str.bytesize == len`

**Comment by eregon** [comment]
Here is it as a PR: https://github.com/ffi/ffi/pull/858

**Comment by eregon** [comment]
I must admit the `str.bytesize == len` condition is a bit strange though, and might not detect all cases.

Maybe we should always warn on `write_string(str, len)`? (and no `\0` until FFI 2)

Probably such usages should become either `write_bytes(str) / write_bytes(str, 0, len)` or `write_string(str.byteslice(0, len))`.
It seems very rare to need to write the prefix of a String and still null-terminate, so I think the `byteslice` usage is fair enough for that case (could also do `write_bytes(str, 0, len); put_char(len, 0)` in user code to be explicit and slightly more efficient).

**Comment by eregon** [comment]
PR for that second approach, which seems more reliable and compatible: https://github.com/ffi/ffi/pull/859

**Comment by larskanis** [comment]
I was afraid to get these kinds of incompatibilities... But #858 or #859 still don't recover compatibility in all cases. Because we can not fix the inconsistencies without breaking compatibility, since we can not guess the intention of the calling code.

Instead I propose to just revert #806 for ffi-1.14.1 and discuss a better strategy afterwards.
@eregon Is this OK for you?

In case of https://github.com/chuckremes/ffi-rzmq/blob/master/lib/ffi-rzmq/message.rb#L126-L135 I'd suggest using `MemoryPointer.from_string` with `autorelease = false` instead. It is more portable than using `LibC.malloc` and has bound checks in addition. That is described in https://github.com/ffi/ffi/wiki/Core-Concepts#using-mallocfree

On the other hand we [recommend since years](https://github.com/ffi/ffi/wiki/Pointers#handing-off-memory-to-external-libraries) to use
```ruby
foo = "a Ruby string"
buffer1 = LibC.malloc foo.size
buffer1.write_string foo
```
which doesn't allow a NUL termination and is plain wrong in case of non ASCII characters or UTF-16 in `foo`.

**Comment by eregon** [comment]
> But #858 or #859 still don't recover compatibility in all cases. 

Which case are you thinking?
It seems only `write_string(str)` changes to null-terminate if we merge #859.

> Because we can not fix the inconsistencies without breaking compatibility, since we can not guess the intention of the calling code.

I think in many cases we can though.
At least the intention of `write_string(str, len)` seems clear, and #859 fixes that to behave the same as before, and a way forward.

For `write_string(str)`, I would argue most usages expect null termination, even though it did not happen before and it might cause subtle bugs.

> Instead I propose to just revert #806 for ffi-1.14.1 and discuss a better strategy afterwards.
> @eregon Is this OK for you?

It's your call, but I think merging #859 and releasing 1.14.1 with it would be a good solution.
Then if `write_string(str)` is problematic too, we can revert it.

The thing is with #859, we still have a clear plan:
* `write_string_length` and `write_string(str, len)` are deprecated
* They can be removed in FFI 2, which is great because it means no confusion possible for these two cases.

If we undo null-termination for `write_string(str)`, then there seems to be no way to migrate or warn anyway, except maybe deprecating the whole method and removing/renaming it, which seems suboptimal.
Null-termination is useful, and it seems to have always been there for `put_string(offset, str)`, so it's only natural it's there for `write_string(str)` (#805).

---

> In case of https://github.com/chuckremes/ffi-rzmq/blob/master/lib/ffi-rzmq/message.rb#L126-L135

How would you pass a pointer to the free() function? `LibC::Free` as it is?

> On the other hand we [recommend since years](https://github.com/ffi/ffi/wiki/Pointers#handing-off-memory-to-external-libraries) to use

That's just bad documentation, it happens, let's remove it or fix it.

**Comment by eregon** [comment]
I noticed my current proposition with #859 is actually quite close to what I mentioned a while ago: https://github.com/ffi/ffi/issues/805#issuecomment-657062681

There I didn't think about a deprecation warning, but I think it's actually a good way to fix the `write_string(str, len)`/`write_string_length` cases.

So with #859, `write_string(*args)` would finally be just `put_string(0, *args)`, like every other `write_` method, making everything consistent.

**Comment by larskanis** [comment]
> Which case are you thinking?
> It seems only `write_string(str)` changes to null-terminate if we merge #859.

Exactly the case the [documentation recommends since years](https://github.com/ffi/ffi/wiki/Pointers#handing-off-memory-to-external-libraries). 

> The thing is with #859, we still have a clear plan:

I don't think so. `get_string` doesn't stop at a NUL character but `get_array_of_string` does. That's the other inconsistency we didn't work on yet.

**Comment by larskanis** [comment]
My fault - `get_string` reads up to the NUL byte. I didn't understand the docs.

**Comment by larskanis** [comment]
> * `write_string_length` and `write_string(str, len)` are deprecated

I think there isn't anything bad in having a method with length limitation and a terminating NUL.

**Comment by larskanis** [comment]
I just released ffi-1.14.1 which reverts the change to `write_string`. See #861.

**Comment by larskanis** [comment]
> In case of https://github.com/chuckremes/ffi-rzmq/blob/master/lib/ffi-rzmq/message.rb#L126-L135 I'd suggest using `MemoryPointer.from_string` with `autorelease = false` instead. It is more portable than using `LibC.malloc` and has bound checks in addition. That is described in https://github.com/ffi/ffi/wiki/Core-Concepts#using-mallocfree

@skaes I played a little bit with this piece of code.

It's possible to just pass the `bytes` string to C and avoid string copying, but that needs additional measurement to avoid GC'ing until the freeing callback is called. It turned out that this additional measurement is more expensive than copying the message to malloc()'ed space (at least with message sizes of around 100 bytes) and directly passing the function pointer to `free()`.

Strictly speaking the direct passing of the function pointer to `free()` is a violation to the function signature of `zmq_free_fn` which has an additional `hint` parameter. But practically all relevant call conventions accept this violation. And it avoids an additional callback into ruby space.

So `write_string(bytes, len)` might be changed to `write_bytes(bytes, 0, len)`, but the rest is best how it is.
